ConcurrentModificationException when late meter filters are added

mabartos opened the pull request:
- Fixes #5489

@lenin-jaganathan You've already mentioned that it should be synchronized, but no action were taken (https://github.com/micrometer-metrics/micrometer/pull/4857/commits/778f400174e54935a5948f17c283713edb912258#r1555315425)

Is it ok to lock it with the `meterMapLock`, or rather use a concurrent collection?

pivotal-cla commented:
@mabartos Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/pivotal?repositoryId=micrometer-metrics/micrometer&pullRequestId=5490)!

[Click here](https://cla.pivotal.io/sync/pivotal?repositoryId=micrometer-metrics/micrometer&pullRequestId=5490) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

pivotal-cla commented:
@mabartos Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/pivotal?repositoryId=micrometer-metrics/micrometer&pullRequestId=5490)!

lenin-jaganathan commented:
I wouldn't recommend a `Collections.synchronizedSet()` if that is what you mean by "concurrent collection". We use `contains` on every interaction with the builder which could cause slowness.

We can use `meterMapLock` (but it was not originally meant for this map) but we should guard all the writes atleast. I also see that this collection is declared as not "Thread safe".

mabartos commented:
@lenin-jaganathan Thanks for the comment.

> I wouldn't recommend a Collections.synchronizedSet() if that is what you mean by "concurrent collection". We use contains on every interaction with the builder which could cause slowness.

Ok.

> We can use meterMapLock (but it was not originally meant for this map) but we should guard all the writes atleast. I also see that this collection is declared as not "Thread safe".

All the writes/removes should be guarded now. 

@lenin-jaganathan Let me know, if there's a better approach. But resolving this bug is crucial for us as it would need to be propagated to Quarkus as well, so it'd be great to prioritize this issue.

mabartos commented:
The CI `build-jdk11` failed due to a test failure related to some timing issue after providing `synchronized()` to `unmarkStaleId()` method. Not sure if it is related to this PR, or just the test is unstable.

shakuzen commented:
I re-ran the jdk11 job and it passed. I'm not sure if the failures were related to the change or not.
Note I'm holding off reviewing the specific changes in this PR and running benchmarks until I can get a concurrency test that reproduces the failure this is fixing. We want to make sure we've really fixed it and that it stays fixed through future changes. I'm out of time to continue working on that today but I'll get back to it tomorrow.

shakuzen commented:
I was finally able to figure out a test that reproduces the CME. It wasn't obvious to me at first, but the CME is being thrown because `preFilterIdToMeterMap` is being modified (a new entry added) while it is being iterated over. The iteration is implicit in the foreach call in `addAll` that's being passed the keySet of `preFilterIdToMeterMap`. With a test that's hopefully reproducing the same issue as the one reported, we can work on the best solution for it. Unfortunately I'm a bit under the weather today, but if I'm feeling well enough tomorrow I'll continue with this work then.

Code:

```java
@JCStressTest
@Outcome(id = "true", expect = Expect.ACCEPTABLE, desc = "No exception")
@Outcome(expect = Expect.FORBIDDEN, desc = "ConcurrentModificationException thrown")
@State
public static class ConfigureLateMeterFilterWithNewMeterRegister {
    MeterRegistry registry = new SimpleMeterRegistry();

    public ConfigureLateMeterFilterWithNewMeterRegister() {
        // need the registry to not be empty
        registry.counter("c1");
    }

    @Actor
    public void actor1() {
        // creates new meter to add to preMap, updating the modCount
        registry.counter("c2");
    }

    @Actor
    public void actor2(Z_Result r) {
        try {
            // adds all the preMap keys to the staleIds, iterating over preMap keys
            registry.config().commonTags("common2", "tag2");
            r.r1 = true;
        } catch (ConcurrentModificationException e) {
            r.r1 = false;
        }
    }
}
```

Results:

```
  Results across all configurations:

  RESULT     SAMPLES     FREQ      EXPECT  DESCRIPTION
   false      57,671    0.19%   Forbidden  ConcurrentModificationException thrown
    true  30,863,033   99.81%  Acceptable  No exception

```

mabartos commented:
@shakuzen After applying these changes, the test results are as follows:

```
[OK] io.micrometer.concurrencytests.MeterRegistryConcurrencyTest.ConfigureLateMeterFilterWithNewMeterRegister

  Results across all configurations:

  RESULT    SAMPLES     FREQ      EXPECT  DESCRIPTION
    true  8,239,104  100.00%  Acceptable  No exception

```

Let me know if I can help somehow.

shakuzen commented on micrometer-core/src/main/java/io/micrometer/core/instrument/MeterRegistry.java:
```diff
@@ -692,7 +695,12 @@ private boolean isStaleId(Id originalId) {
     private boolean unmarkStaleId(Id originalId) {
-        return !stalePreFilterIds.isEmpty() && stalePreFilterIds.remove(originalId);
+        if (!stalePreFilterIds.isEmpty()) {
+            synchronized (meterMapLock) {
```
This isn't needed for fixing the CME and it's not clear to me what scenario this improves (it also doesn't seem to hurt anything either, though). I am leaning toward not including this change unless we have a concrete idea on that.

mabartos replied:
@shakuzen You're right it's not needed for fixing the actual issue, but it was mainly provided after @lenin-jaganathan comment that all writes(write ops) should be covered as we also stated in the comment of the collection. 

But I'll update it.

shakuzen reviewed (commented):
I left a comment about one of the changes. I have some polish to make for the concurrency test, and we will want to rebase on the `1.13.x` branch so we can release the fix in the next maintenance release, but I can take care of both of these things before merging.

mabartos commented:
@shakuzen I updated the code and squashed commits.

Created backport for 1.13.x branch: https://github.com/micrometer-metrics/micrometer/pull/5496

Not sure how you handle backports, so let me know if I should manage it differently. 

We need to have the new patch 1.13.x version ASAP to include it in Quarkus and other project. 

Thanks for your understanding.

brunobat commented:
> 
> We need to have the new patch 1.13.x version ASAP to include it in Quarkus and other project.
> 

Not sure they can do that...
If there are only fixes in the next patch release we can roll forward and backport to the Quarkus LTS branch.

shakuzen commented:
> Not sure how you handle backports, so let me know if I should manage it differently.

No worries. We merge to the oldest applicable maintenance branch and then merge forward from there. I'll take care of it from here.

mabartos commented:
> No worries. We merge to the oldest applicable maintenance branch and then merge forward from there. I'll take care of it from here.

@shakuzen Ok, thank you!
